refactor: extract aggregator actors

jitsedesmet opened the pull request:
This PR aims to extract the aggregator functionality into multiple actors. We already have aggregator components, these can be replaced by actors.

jitsedesmet commented:
I already came across one thorn, the wildcard count was handled as an edge case (because it is). To extract the aggregators to actors, we need to generalize to that level. That's why I chose to generalize at top level and drop the `SyncAggregator`.

jitsedesmet commented:
Pleased to announce. Extraction of aggregators is done. (it is ugly but spec and integration tests pass so I'm happy)

rubensworks commented:
🥳

rubensworks commented on packages/bus-expression-evaluator-aggregate/lib/ActorExpressionEvaluatorAggregate.ts:
```diff
@@ -0,0 +1,47 @@
+ * @see IActionExpressionEvaluatorAggregate
+ * @see IActorExpressionEvaluatorAggregateOutput
+ */
+export abstract class ActorExpressionEvaluatorAggregate extends Actor<
```
Perhaps `ActorBindingsAggregatorFactory` or `ActorExpressionAggregatorFactory` might be a more convenient name.

jitsedesmet replied:
ActorBindingsAggregatorFactory :)

rubensworks commented on packages/actor-expression-evaluator-aggregate-group-concat/lib/ActorExpressionEvaluatorAggregateGroupConcat.ts:
```diff
@@ -0,0 +1,66 @@
+  }
+}
+
+class GroupConcatAggregator extends AggregateEvaluator {
```
Let's move each aggregator to a separate file.

jitsedesmet commented:
I think this PR should be ready. I have a few things to point out:

This PR removes the sync evaluator because it results in inconveniences and a code duplication and overall is hard to maintain (especially because Comunica never uses it).

Since I removed the sync evaluator, a lot of code could be simplified, this PR already does that partially. You will see some remaining ugly code in the partial-evaluator package, please ignore this package ugliness in this PR, and focus on the other packages. The next PR that I make will be a transformation of the EE package itself. The PR will prepare the extraction of the function into actors. (I expect this to be a rewrite of core sparqlee classes, that's why you shouldn't bother with the current state). 

For the other packages: the main contribution of this PR is the bussification of the sparqlee aggregators. These packages should be thoroughly reviewed. (If names need to change, please tell me in one go because that takes a fair amount of work).

Additionally, I removed extension functions for now. I don't know if we want them or not, we should have a discussion surrounding whether we want them and how to implement them. In case we want them, I suggest reimplementing them in the `function-bussification`-PR.

rubensworks commented on packages/actor-bindings-aggregator-factory-average/README.md:
```diff
@@ -0,0 +1,41 @@
+
+### Config Parameters
+
+TODO: fill in parameters (this section can be removed if there are none)
```
TODO (possible also in other READMEs)

jitsedesmet replied:
The bus readme contains this line:
> Actors extending [`ActorBindingsAggregatorFactory`](TODO:jsdoc_url) are automatically subscribed to this bus.


What should be the link?

jitsedesmet replied:
(It seems to be death for other busses?)

rubensworks replied:
It should refer to the class location here: https://comunica.github.io/comunica/

rubensworks commented on packages/bus-bindings-aggeregator-factory/lib/ActorBindingsAggregatorFactory.ts:
```diff
@@ -0,0 +1,40 @@
+ * A comunica actor for creating Binding-Aggregators.
+ *
+ * Actor types:
+ * * Input:  IActionExpressionEvaluatorAggregate:      TODO: fill in.
```
TODO

rubensworks commented on packages/bus-bindings-aggeregator-factory/lib/ActorBindingsAggregatorFactory.ts:
```diff
@@ -0,0 +1,40 @@
+ * Actor types:
+ * * Input:  IActionExpressionEvaluatorAggregate:      TODO: fill in.
+ * * Test:   <none>
+ * * Output: IActorExpressionEvaluatorAggregateOutput: TODO: fill in.
```
TODO

rubensworks commented on packages/bus-query-operation/lib/ActorQueryOperation.ts:
```diff
@@ -14,14 +13,6 @@ import type * as RDF from '@rdfjs/types';
- *
- * @type {number}
- */
-let bnodeCounter = 0;
```
The fact that we can remove all of this stuff here is a sign we're on the right track :-)

jitsedesmet replied:
I just moved it somewhere else though :eyes:

rubensworks replied:
The logic is fine, but the location was not :-)

rubensworks commented on packages/types/lib/ExpressionEvaluator.ts:
```diff
@@ -0,0 +1,20 @@
+import type { Algebra as Alg } from 'sparqlalgebrajs';
+import type { IActionContext } from './IActionContext';
+
+export interface IBindingAggregator {
```
Could you add some jsdocs to this file?

rubensworks reviewed (changes requested):
I like this very much, nice abstraction work @jitsedesmet!

Just some minor comments here and there. We should also make sure that all tests pass with full coverage before merging.

jitsedesmet commented on engines/query-sparql/test/QuerySparql-test.ts:
```diff
@@ -341,7 +341,7 @@ describe('System test: QuerySparql', () => {
       });
 
-      describe('extension function', () => {
+      describe.skip('extension function', () => {
```
I refer to: https://github.com/comunica/comunica/pull/1269#issuecomment-1752033721

jitsedesmet commented on engines/config-query-sparql/config/expression-evaluator/actors.json:
```diff
@@ -0,0 +1,15 @@
+    "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/config-query-sparql/^2.0.0/components/context.jsonld"
+  ],
+  "import": [
+    "ccqs:config/expression-evaluator/aggregators/average.json",
```
Okey :D 
How about this line: 
> Each sub-directory represents a bus

It would not hold for the `expression-evaluator` directory? (I still need the `evaluators` file)

rubensworks replied:
> It would not hold for the expression-evaluator directory? (I still need the evaluators file)

Indeed, our exp-eval is not a bus, but just a component, so that's a bit of a special one.
So having a dedicated dir for that does make sense.
However, I do wonder if we should consider bus-ifying that one as well.

jitsedesmet replied:
I don't think that would make a lot of sense, because we decided to have our function-actors use the helpers provided by the EE (think OverloadTree).

jitsedesmet replied:
> So having a dedicated dir for that does make sense.

Should I edit the rules described in:  https://github.com/comunica/comunica/tree/master/engines/config-query-sparql/config#directory-structure

jitsedesmet replied:
(That being said, I did already make an interface for the EE :) )

jitsedesmet commented on packages/bus-bindings-aggeregator-factory/lib/index.ts:
```diff
@@ -0,0 +1 @@
+export * from './ActorBindingsAggregatorFactory';
```
Yes! Because I make these mistakes, I will try and make a script for the extraction of the functions.

jitsedesmet commented on packages/actor-bindings-aggregator-factory-average/lib/AverageAggregator.ts:
```diff
@@ -0,0 +1,52 @@
+
+export class AverageAggregator extends AggregateEvaluator implements IBindingAggregator {
+  // This will eventually be a mediator call.
+  private readonly summer = regularFunctions[C.RegularOperator.ADDITION];
+  private readonly divider = regularFunctions[C.RegularOperator.DIVISION];
```
Instead of accessing these directly, I think it should be nicer to ask the function to either an EE instance, or ask for a function instance of the EE-Factory. (These things should be considered when functions have been extracted (and they will be considered because extraction will kill this code))

jitsedesmet replied:
(this can be ignored, I just wanted to clarify this)

rubensworks commented on engines/query-sparql-rdfjs/package.json:
```diff
@@ -124,6 +124,7 @@
     "@comunica/bus-query-operation": "^2.9.0",
     "@comunica/config-query-sparql": "^2.7.0",
     "@comunica/core": "^2.8.2",
+    "@comunica/expression-evaluator": "^2.9.0",
```
I guess the bindings-aggregator-factory deps are missing here?

rubensworks commented on packages/actor-query-operation-extend/lib/ActorQueryOperationExtend.ts:
```diff
@@ -73,3 +75,7 @@ export class ActorQueryOperationExtend extends ActorQueryOperationTypedMediated<
 }
+
+interface IActorQueryOperationExtendArgs extends IActorQueryOperationTypedMediatedArgs {
+  expressionEvaluatorFactory: ExpressionEvaluatorFactory;
```
Just realized that the READMEs of this actor (and probably also other actors) will need to modified to mention this as config param.

rubensworks commented on packages/types/lib/ExpressionEvaluator.ts:
```diff
@@ -0,0 +1,73 @@
+ */
+export interface IBindingAggregator {
+  /**
+   * Registers bindings to the aggregator. Each binding you put has the ability to change te result of aggregation.
```
```suggestion
   * Registers bindings to the aggregator. Each binding you put has the ability to change the aggregation result.
```

rubensworks commented on packages/types/lib/ExpressionEvaluator.ts:
```diff
@@ -0,0 +1,73 @@
+import type { IActionContext } from './IActionContext';
+
+/**
+ * Interface that performs a specific aggregation.
```
```suggestion
 * Instances of this interface perform a specific aggregation of bindings.
```

rubensworks commented on packages/types/lib/ExpressionEvaluator.ts:
```diff
@@ -0,0 +1,73 @@
+ * Interface that performs a specific aggregation.
+ * You can put bindings and when all bindings have been put, request the result.
+ */
+export interface IBindingAggregator {
```
```suggestion
export interface IBindingsAggregator {
```
(to be aligned with the bus name)

rubensworks commented on packages/types/lib/ExpressionEvaluator.ts:
```diff
@@ -0,0 +1,73 @@
+
+/**
+ * A factory able to create an expression evaluator and related types.
+ * It can be used as a way to group mediators in a single class, ready for later use.
```
This sentence can be removed.

rubensworks commented on packages/types/lib/ExpressionEvaluator.ts:
```diff
@@ -0,0 +1,73 @@
+}
+
+/**
+ * A factory able to create an expression evaluator and related types.
```
```suggestion
 * A factory able to create objects for handling expressions.
```

rubensworks commented on packages/types/lib/ExpressionEvaluator.ts:
```diff
@@ -0,0 +1,73 @@
+}
+
+/**
+ * An RDF evaluator.
```
```suggestion
 * An evaluator for RDF expressions.
```

rubensworks commented on packages/types/lib/ExpressionEvaluator.ts:
```diff
@@ -0,0 +1,73 @@
+  /**
+   * The factory that created this Expression Evaluator.
+   */
+  factory: IExpressionEvaluatorFactory;
```
Hmm, this feels wrong. Is it really necessary to have this here?

jitsedesmet commented on packages/actor-bindings-aggregator-factory-count/test/util.ts:
```diff
@@ -0,0 +1,28 @@
+import { BindingsFactory } from '@comunica/bindings-factory';
```
These util files should be extracted and moved to the comunica/jest package

jitsedesmet commented:
There seems to be a failing test in `packages/actor-rdf-parse-html-script/test/ActorRdfParseHtmlScript-test.ts`
namely `ActorRdfParseHtml › An ActorRdfParseHtmlScript instance › run › the html listener › should handle two jsonld script tags in graphs`, whats weird is that even when I run this test on master, it fails.

jitsedesmet commented:
For some reason, this commit: https://github.com/comunica/comunica/pull/1269/commits/56b378336b16319344d46896b47e241706c9ffb5 breaks a test in `packages/actor-rdf-parse-html-script/test/ActorRdfParseHtmlScript-test.ts`
namely `ActorRdfParseHtml › An ActorRdfParseHtmlScript instance › run › the html listener › should handle two jsonld script tags in graphs`, I do not remotly tough that file though. @rubensworks do you have an idea what could be the reason for this failing test (I already looked for a few hours and maybe you see what the file has in common?)

rubensworks commented:
@jitsedesmet Not sure where you're looking exactly, but if I look at the CI output, all tests pass, and I see the following depcheck error:

> Error: Extra dependencies: rdf-data-factory in @comunica/actor-bindings-aggregator-factory-group-concat

rubensworks commented on packages/actor-query-operation-leftjoin/README.md:
```diff
@@ -41,3 +41,4 @@ After installing, this package can be added to your engine's configuration as fo
 
 * `mediatorQueryOperation`: A mediator over the [Query Operation bus](https://github.com/comunica/comunica/tree/master/packages/bus-query-operation).
 * `mediatorJoin`: A mediator over the [RDF Join bus](https://github.com/comunica/comunica/tree/master/packages/bus-rdf-join).
+* `expressionEvaluatorFactory`: A factory to create an [Expression Evaluator](https://github.com/comunica/comunica/tree/master/packages/expression-evaluator);
```
Could you also add this param to the example above in this file? (also applies to the other readme's using expressions)

rubensworks commented on packages/jest/lib/ExpressionEvaluator.ts:
```diff
@@ -0,0 +1,61 @@
+import { BindingsFactory } from '@comunica/bindings-factory';
```
Can we name this file `expressions.ts`? Otherwise, it feels like this class exposes the `ExpressionEvaluator` class.
